h2o-core/src/main/java/water/rapids/ast/prims/assign/AstRectangleAssign.java

```diff
@@ -196,66 +196,36 @@ void mapChunkSlice(Chunk[] cs, int chkOffset) {
     }
   }
 
-  // Assign a NON-STRING SCALAR over some dst rows; optimize for all rows
-  private void assign_frame_scalar(Frame dst, int[] cols, AstNumList rows, double src, Session ses) {
-
-    // Handle fast small case
+  // Assign a SCALAR over some dst rows; optimize for all rows
+  private void assign_frame_scalar(Frame dst, int[] cols, AstNumList rows, Object src, Session ses) {
     long nrows = rows.cnt();
-    if (nrows == 1) {
-      Vec[] vecs = ses.copyOnWrite(dst, cols);
-      long drow = (long) rows._bases[0];
-      for (int col : cols)
-        vecs[col].set(drow, src);
-      return;
-    }
 
-    // Bulk assign constant (probably zero) over a frame.  Directly set
+    // Bulk assign a numeric constant (probably zero) over a frame.  Directly set
     // columns: Copy-On-Write optimization happens here on the apply() exit.
-    if (dst.numRows() == nrows && rows.isDense()) {
+    // Note: this skips "scalar to Vec" compatibility check because the whole Vec is overwritten
+    if (dst.numRows() == nrows && rows.isDense() && (src instanceof Number)) {
       Vec anyVec = dst.anyVec();
       assert anyVec != null;  // if anyVec was null, then dst.numRows() would have been 0
-      Vec vsrc = anyVec.makeCon(src);
+      Vec vsrc = anyVec.makeCon((double) src);
       for (int col : cols)
         dst.replace(col, vsrc);
       if (dst._key != null) DKV.put(dst);
       return;
     }
 
```

Comment:
NAs and Strings are handled as the "large" case in an MRTask. I don't think having a special handling here really provides any significant benefits.